lib/yabeda/prometheus/exporter.rb

```diff
@@ -18,26 +18,51 @@ def call(env)
           @app.call(env)
         end
 
-        def start_metrics_server!(**rack_app_options)
+        def start_metrics_server!(start_in_thread: true, **rack_app_options)
+          if start_in_thread
+            start_server_in_thread!(**rack_app_options)
+          else
+            start_server_in_process!(**rack_app_options)
+          end
+        end
+
+        def start_server_in_thread!(**rack_app_options)
           Thread.new do
-            default_port = ENV.fetch("PORT", 9394)
-            ::Rack::Handler::WEBrick.run(
-              rack_app(**rack_app_options),
-              Host: ENV["PROMETHEUS_EXPORTER_BIND"] || "0.0.0.0",
-              Port: ENV.fetch("PROMETHEUS_EXPORTER_PORT", default_port),
-              AccessLog: [],
-            )
+            start_app(**rack_app_options)
+          end
+        end
+
+        def start_server_in_process!(**rack_app_options)
+          pid = Process.fork do
+            # re-configure yabeda since we're in a new process
+            Yabeda.configure!
+            start_app(**rack_app_options)
           end
+          Process.detach(pid) if pid
         end
 
-        def rack_app(exporter = self, path: "/metrics", logger: Logger.new(IO::NULL))
+        def rack_app(exporter = self, path: "/metrics", logger: Logger.new(IO::NULL), use_deflater: false)
           ::Rack::Builder.new do
+            use ::Rack::Deflater if use_deflater
             use ::Rack::CommonLogger, logger
             use ::Rack::ShowExceptions
             use exporter, path: path
             run NOT_FOUND_HANDLER
           end
         end
+
+        def start_app(raise_start_error: true, **rack_app_options)
+          default_port = ENV.fetch("PORT", 9394)
+          ::Rack::Handler::WEBrick.run(
+            rack_app(**rack_app_options),
+            Host: ENV["PROMETHEUS_EXPORTER_BIND"] || "0.0.0.0",
+            Port: ENV.fetch("PROMETHEUS_EXPORTER_PORT", default_port),
+            AccessLog: [],
+          )
+        rescue Errno::EADDRINUSE
+          puts "Yabeda prometheus exporter: Failed to start, server might be started by another process"
+          raise if raise_start_error
```

Comment:
To be honest, this looks weird and feels wrong to me (see review summary for the details)

Reply:
yeah that's fair 😅 can remove this change